File: Bruteforce/LC/TOP/JS-version/151.js
var reverseWords = function (s) {
  let n = s.length
  // Convert string to array for manipulation
  let arr = s.split('').reverse()

  console.log(arr)
  let finalIndex = 0

  for (let i = 0; i < n; i++) {
    if (arr[i] !== ' ') {
      // Add space between words
      if (finalIndex !== 0) {
        arr[finalIndex++] = ' '
      }

      let wordStart = finalIndex
      let j = i

      while (j < n && arr[j] !== ' ') {
        arr[finalIndex++] = arr[j++]
      }

      // Reverse the current word
      let wordEnd = finalIndex
      let left = wordStart
      let right = wordEnd - 1
      while (left < right) {
        ;[arr[left], arr[right]] = [arr[right], arr[left]]
        left++
        right--
      }

      i = j
    }
  }

  // Resize array and join to string
  return arr.slice(0, finalIndex).join('')
}

// Test
const s = 'the sky is blue'
console.log(reverseWords(s))

File: Bruteforce/LC/TOP/151.cpp
/*
Input: s = "the sky is blue"
Output: "blue is sky the"
*/

#include <iostream>
#include <bits/stdc++.h>
using namespace std;

string reverseWords(string s)
{
    int n = s.size();
    reverse(s.begin(), s.end());

    int finalIndex = 0;

    for (int i = 0; i < n; ++i)
    {
        if (s[i] != ' ')
        {
            // space
            if (finalIndex != 0) s[finalIndex++] = ' ';
            int wordStart = finalIndex;
            int j = i;

            while (j < n && s[j] != ' ')
            {
                s[finalIndex++] = s[j++];
            }

            // wordstart not word start -1 as wordstart start from 0 (s.begin()+0 fo ex)
            reverse(s.begin() + wordStart, s.begin() + finalIndex);
            // j is space;
            i = j;
        }
    }

    s.resize(finalIndex);
    return s;
}

int main()
{
    string s = "the sky is blue";
    s = reverseWords(s);
    cout << s;
    return 0;
}